proc/switcher/switch.go

```diff
@@ -1,169 +1,78 @@
+	vals   []zed.Value
 }
 
-// gather ensures that we have a request for each active downstream.
-func (s *Switcher) gather() {
-	var nready int
-	for _, r := range s.requests {
-		if r != nil {
-			nready++
-		}
-	}
-	for nready < s.n {
-		req := <-s.requestsCh
-		if req.ch == nil {
-			s.n--
-			continue
-		}
-		nready++
-		if s.requests[req.i] != nil {
-			panic("Switcher bug")
-		}
-		s.requests[req.i] = req.ch
+func New(pctx *proc.Context, parent proc.Interface) *Selector {
+	router := proc.NewRouter(pctx, parent)
+	s := &Selector{
+		Router: router,
 	}
+	router.Link(s)
+	return s
 }
 
-func (s *Switcher) run() {
-	records := make([][]zed.Value, s.n)
-	results := make([]proc.Result, s.n)
-	for {
-		s.gather()
-		if s.n == 0 {
-			break
-		}
-		batch, err := s.parent.Pull()
-		if batch == nil || err != nil {
-			s.sendEOS(proc.Result{batch, err})
-			continue
-		}
-		ectx := batch.Context()
-		vals := batch.Values()
-		for i := range vals {
-			this := &vals[i]
-			for j, f := range s.filters {
-				val := f.Eval(ectx, this)
-				if val.IsError() {
-					// XXX should use structured here to wrap
-					// the input value with the error
-					records[j] = append(records[j], *val)
-					break
-				}
-				if val.Type == zed.TypeBool && val.Bytes != nil && zed.IsTrue(val.Bytes) {
-					records[j] = append(records[j], vals[i])
-					break
-				}
+func (s *Selector) AddCase(f expr.Evaluator) proc.Interface {
+	route := s.Router.AddRoute()
+	s.cases = append(s.cases, &switchCase{filter: f, route: route})
+	return route
+}
 
+func (s *Selector) Forward(router *proc.Router, batch zbuf.Batch) error {
+	ectx := batch.Context()
+	vals := batch.Values()
+	for i := range vals {
+		this := &vals[i]
+		for _, c := range s.cases {
+			val := c.filter.Eval(ectx, this)
+			if val.IsMissing() {
+				continue
 			}
-		}
-		for i := range records {
-			if records[i] != nil {
-				results[i] = proc.Result{Batch: zbuf.NewArray(records[i])}
-				records[i] = records[i][:0]
+			if val.IsError() {
+				// XXX should use structured here to wrap
+				// the input value with the error
+				c.vals = append(c.vals, *val)
+				continue
+				//XXX don't break here?
+				//break
+			}
+			if val.Type == zed.TypeBool && val.Bytes != nil && zed.IsTrue(val.Bytes) {
+				c.vals = append(c.vals, *this)
+				break
 			}
-		}
-		s.send(results)
-		batch.Unref()
-	}
-	s.parent.Done()
-}
-
-func (s *Switcher) sendEOS(result proc.Result) {
-	for i := range s.requests {
-		if s.requests[i] != nil {
-			s.requests[i] <- result
-			s.requests[i] = nil
 		}
 	}
-}
-
-func (s *Switcher) send(results []proc.Result) {
-	for i, ch := range s.requests {
-		if results[i].Batch != nil {
-			results[i].Batch.Ref()
-			ch <- results[i]
-			s.requests[i] = nil
-			results[i].Batch = nil
+	// Send each case that has vals from this batch.
+	// We have vals that point into the current batch so we
+	// ref the batch for each outgoing new batch, then unref
+	// the batch once after the loop.
```

Comment:
This comment reads like this function calls Unref but it's actually the router that calls it.
```suggestion
	// We have vals that point into the current batch so we
	// ref the batch for each outgoing new batch.
```